tornado/iostream.py

```diff
@@ -1023,7 +1023,8 @@ def read_from_fd(self):
     def write_to_fd(self, data):
         return self.socket.send(data)
 
-    def connect(self, address, callback=None, server_hostname=None):
+    def connect(self, address, callback=None, server_hostname=None,
+                source_ip=None):
```

Comment:
I don't think this is the best place to add source_ip. It forces you to add it to all the IOStream classes, even though it's only relevant to TCP. And as you see from the test failures you missed some spots:

https://ci.appveyor.com/project/bdarnell/tornado/build/1.0.493/job/gr2o8lgoejlosgop

Instead I think that TCPClient._create_stream should do the work of calling "bind" on the socket before wrapping it in an IOStream.